### PR TITLE
[JENKINS-67237] BuildTrigger waits until the dependency graph has been updated - 2nd attempt

### PR DESCRIPTION
**Pull request by Si-So**
See [JENKINS-67237](https://issues.jenkins-ci.org/browse/JENKINS-67237).

The first [PR](https://github.com/jenkinsci/jenkins/pull/6101) caused this [regression](https://issues.jenkins.io/browse/JENKINS-68189). This PR is the 2nd attempt at solving [JENKINS-67237](https://issues.jenkins-ci.org/browse/JENKINS-67237) without causing a regression. 

### Proposed changelog entries

* When triggering a new build while the build graph is currently being re-computed, jenkins waits for the re-computation to finish. This gurantees that recently updated build triggers are executed.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@[res0nance](https://github.com/res0nance)
@[timja](https://github.com/timja)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by Si-So** [comment]
I created a TC to detect the [regression](https://issues.jenkins.io/browse/JENKINS-68189). I expect that only the new TC will fail during the currently running build. 

I have also prepared a fix for the regression that I will push once the currently running build is done.

**Comment by Si-So** [comment]
As predicted only the newly added TC failed.

I have now pushed a fix for the regression and expect no failing unit tests during this run.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (res0nance) [inline core/src/main/java/jenkins/model/Jenkins.java:4975]: Because of how this works, potentially a call to `rebuildDependencyGraphAsync` will block which defeats the original purpose of the function. This needs a rethink

Comment (Si-So) [reply core/src/main/java/jenkins/model/Jenkins.java:4976]: Thanks yet again for your feedback. I just pushed a new commit. 

I also noticed another bug: getFutureDependencyGraph does not return the most recent dependency graph if the most recent dependency graph has been calculated synchronously. This is problematic because BuildTrigger.execute assumes that getFutureDependencyGraph will always return the most recent dependency graph. To fix this problem calculatingFutureDependencyGraph is now set to null in scheduleCalculationOfFutureDependencyGraph after the calculation has finished.

**Review by res0nance — APPROVED** [review]
LGTM

**Review by basil — APPROVED** [review]
I pushed a minor commit to clean up the formatting. Thanks for taking care of this!

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by basil** [comment]
The `skip-changelog` label was added by mistake.

**Pull request merged**
